crates/bevy_ecs/src/resource.rs

```diff
@@ -72,4 +86,130 @@ pub use bevy_ecs_macros::Resource;
     label = "invalid `Resource`",
     note = "consider annotating `{Self}` with `#[derive(Resource)]`"
 )]
-pub trait Resource: Send + Sync + 'static {}
+pub trait Resource: Component<Mutability = Mutable> {}
+
+/// The `on_add` component hook that maintains the uniqueness property of a resource.
+pub fn resource_on_add_hook(mut deferred_world: DeferredWorld, context: HookContext) {
+    let world = deferred_world.deref();
```

Comment:
Why is this necessary?  Shouldn't autoderef mean you can write `deferred_world.resource_entities`?